gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -358,8 +360,47 @@ MacroBuiltin::include_str_handler (Location invoc_locus,
 
   std::vector<uint8_t> bytes = load_file_bytes (target_filename.c_str ());
 
-  /* FIXME: Enforce that the file contents are valid UTF-8.  */
-  std::string str ((const char *) &bytes[0], bytes.size ());
+  /* unoptimized check for valid UTF-8 */
+  int expect_single = 0;
+  for (uint8_t b : bytes)
+    {
+      if (expect_single)
+	{
+	  if ((b & 0xC0) != 0x80)
+	    {
+	      /* character was truncated, exit with expect_single != 0 */
+	      break;
+	    }
+	  expect_single--;
+	}
+      else if (b & 0x80)
+	{
+	  if (b >= 0xF8)
+	    {
+	      /* more than 4 leading 1s */
+	      expect_single = 1;
+	      break;
+	    }
+	  expect_single = __builtin_clz (~b) - 1;
+	  if (!expect_single)
+	    {
+	      /* only 1 leading 1 */
+	      expect_single = 1;
+	      break;
+	    }
+	}
+    }
+
+  std::string str;
+  if (expect_single)
+    {
+      rust_error_at (invoc_locus, "%s wasn't a utf-8 file",
+		     target_filename.c_str ());
+    }
+  else
+    {
+      str = std::string ((const char *) &bytes[0], bytes.size ());
+    }
```

Comment:
```suggestion
  if (expect_single)
      rust_error_at (invoc_locus, "%s wasn't a utf-8 file",
		     target_filename.c_str ());
  else
      str = std::string ((const char *) &bytes[0], bytes.size ());
```
Same tiny styling nit